teps/XXXX-step-workspaces.md

```diff
@@ -0,0 +1,254 @@
+- Validation of the configuration of a WorkspaceBinding
+- Overriding configuration of a WorkspaceBinding
+
+## Requirements
+
+- A Task author should be able to limit access to a `Workspace` to only those `Steps` that actually
+require the contents of that `Workspace`. By doing so they should be able to limit the running code
+that has access to those contents as well.
+- A Task author should be able to use a Workspace from a `Sidecar`.
+
+## Proposal
+
+### Add `workspaces` to `Steps`
+
+1. Add a `workspaces` list to `Steps`.
+2. Allow `workspaces` from the `Task` to be explicitly named in that list like this:
+
+    ```yaml
+    workspaces:
+    - name: my-sensitive-workspace
+    steps:
+    - name: foo
+      workspaces:
+      - name: my-sensitive-workspace
+    ```
+3. A `workspace` named in a `Step` is not given to other `Steps` unless they too have the
+`workspace` in their own `workspaces` list.
+
+Example YAML:
+
+```yaml
+# task spec
+spec:
+  workspaces:
+  - name: git-ssh-credentials
+    mountPath: /root/.ssh
+  steps:
+  - name: clone-repo
+    image: alpine/git:v2.26.2
+    workspaces:
+    - name: git-ssh-credentials
+    script: |
+      git clone $(params.repo-url) /workspace/source
+  - name: run-unit-tests
+    script: |
+      cd /workspace/source
+      go test ./...
+```
+
+In the above example only the `clone-repo` `Step` will receive access to the `git-ssh-credentials`
+`Workspace`. The `run-unit-tests` `Step` will not receive access to the Workspace volume. Importantly
+this also means that the user-supplied code does not have access to the credential files. Compromising
+the code for the unit tests does not also compromise the SSH credentials.
+
+### Add `workspaces` to `Sidecars`
+
+1. Add a `workspaces` list to `Sidecars`.
+2. Allow `workspaces` from the `Task` to be explicitly named in that list like this:
+
+    ```yaml
+    workspaces:
+    - name: my-workspace
+    sidecars:
+    - name: watch-workspace
+      workspaces:
+      - name: my-workspace
+    ```
+
+The Sidecar `watch-workspace` will now have access to the `my-workspace` Workspace.
+
+### User Stories
+
+#### Story 1
+
+An author of the [`buildpacks-phases`](https://github.com/tektoncd/catalog/blob/master/task/buildpacks-phases/0.1/buildpacks-phases.yaml)
+Catalog task may want to rewrite the Task to reduce the possible blast radius of
+running untrusted images by limiting exposure of Docker credentials to only
+the Step which needs them to push images.
+
+In the buildpacks-phases Task there are 7 Steps and only 1 appears to need docker
+credentials. The are 6 other Steps that will currently receive creds-init docker
+credentials, run different images with different scripts and programs that could
+each be a vector to compromise those credentials.
+
+#### Story 2
+
+As a Task author I want to have a blessed method of sharing a Workspace between Steps and Sidecars
+so that I can pass configuration, messages, data, and logs between them.
```

Comment:
do we have a specific use case where someone wants to do this?

Reply:
Ah, good catch!  Reading back it looks like Stories 2, 3 and 4 are all very woolly 🐑 so I've rewritten to make them concrete 🏙️ .